core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -237,601 +220,1206 @@ private SqlQueryScheduler(
-        Map<PlanFragmentId, Optional<int[]>> bucketToPartitionMap = createBucketToPartitionMap(plan, partitioningCache);
-        Map<PlanFragmentId, OutputBufferManager> outputBufferManagers = createOutputBufferManagers(stages.values(), bucketToPartitionMap);
-        ImmutableList.Builder<PipelinedStageExecution> executionsBuilder = ImmutableList.builder();
-        ImmutableMap.Builder<StageId, StageScheduler> schedulersBuilder = ImmutableMap.builder();
-        PipelinedStageExecution rootExecution = createPipelinedExecution(
-                plan,
-                outputBufferManagers,
-                bucketToPartitionMap,
-                partitioningCache,
-                taskLifecycleListener,
-                executionsBuilder,
-                schedulersBuilder);
-        List<PipelinedStageExecution> executions = executionsBuilder.build();
-        ExecutionSchedule executionSchedule = executionPolicy.createExecutionSchedule(executions);
-
-        rootExecution.addStateChangeListener(state -> {
-            if (state == FINISHED) {
-                queryStateMachine.transitionToFinishing();
-            }
-            else if (state == CANCELED) {
-                // output stage was canceled
-                queryStateMachine.transitionToCanceled();
-            }
-        });
-
-        Set<StageId> finishedStages = newConcurrentHashSet();
-        for (PipelinedStageExecution execution : executions) {
-            execution.addStateChangeListener(state -> {
-                if (queryStateMachine.isDone()) {
-                    return;
-                }
-                if (!state.canScheduleMoreTasks()) {
-                    dynamicFilterService.stageCannotScheduleMoreTasks(execution.getStageId(), execution.getAllTasks().size());
-                }
-                if (state == FAILED) {
-                    RuntimeException failureCause = execution.getFailureCause()
-                            .map(ExecutionFailureInfo::toException)
-                            .orElseGet(() -> new VerifyException(format("stage execution for stage %s is failed by failure cause is not present", execution.getStageId())));
-                    stages.get(execution.getFragment().getId()).transitionToFailed(failureCause);
-                    queryStateMachine.transitionToFailed(failureCause);
-                }
-                else if (state == ABORTED) {
-                    // this should never happen, since abort can only be triggered in query clean up after the query is finished
-                    queryStateMachine.transitionToFailed(new TrinoException(GENERIC_INTERNAL_ERROR, "Query stage was aborted"));
+        private final QueryStateMachine queryStateMachine;
+        private final Map<StageId, SqlStageExecution> allStages;
+        private final List<SqlStageExecution> coordinatorStagesInTopologicalOrder;
+        private final List<SqlStageExecution> distributedStagesInTopologicalOrder;
```

Comment:
Is the order important here? If not drop from name. If yes maybe add a short comment why.